## Win probability modeling

## EDA

In [1]:
import pandas as pd

df = pd.read_csv('play_by_play_2019-2023.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 869804 entries, 0 to 869803
Data columns (total 27 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   id                  869804 non-null  int64  
 1   offense             869804 non-null  object 
 2   offense_conference  867279 non-null  object 
 3   defense             869804 non-null  object 
 4   defense_conference  867100 non-null  object 
 5   home                869804 non-null  object 
 6   away                869804 non-null  object 
 7   offense_score       869804 non-null  int64  
 8   defense_score       869804 non-null  int64  
 9   game_id             869804 non-null  int64  
 10  drive_id            869804 non-null  int64  
 11  drive_number        869804 non-null  int64  
 12  play_number         869804 non-null  int64  
 13  period              869804 non-null  int64  
 14  clock               869804 non-null  object 
 15  offense_timeouts    867807 non-nul

In [3]:
df['game_id'].nunique()

4966

In [5]:
df.describe().round(2).T

count          mean           std           min  \
id                869804.0  4.010826e+17  1.046829e+16 -5.700000e+02   
offense_score     869804.0  1.349000e+01  1.270000e+01 -9.300000e+01   
defense_score     869804.0  1.399000e+01  1.287000e+01 -6.300000e+01   
game_id           869804.0  4.013557e+08  1.486928e+05  4.011107e+08   
drive_id          869804.0  2.660722e+10  1.748596e+10 -8.200000e+01   
drive_number      869804.0  1.274000e+01  7.480000e+00  1.000000e+00   
play_number       869804.0  5.080000e+00  3.560000e+00  1.000000e+00   
period            869804.0  2.500000e+00  1.110000e+00  1.000000e+00   
offense_timeouts  867807.0  2.530000e+00  9.600000e-01 -1.200000e+01   
defense_timeouts  867807.0  2.540000e+00  9.600000e-01 -1.100000e+01   
yard_line         869804.0  5.051000e+01  2.408000e+01  0.000000e+00   
yards_to_goal     869804.0  5.173000e+01  2.403000e+01 -2.200000e+01   
down              869804.0  1.910000e+00  1.070000e+00  0.000000e+00   
distance          869804.0  8.310000e+00  4.310000e+00 -2.300000e+01   
yards_gained      869804.0  5.700000e+00  1.086000e+01 -1.000000e+02   
ppa               659157.0  1.800000e-01  1.330000e+00 -1.279000e+01   

                           25%           50%           75%           max  
id                4.012473e+17  4.014041e+17  4.015203e+17  4.015539e+17  
offense_score     3.000000e+00  1.000000e+01  2.100000e+01  1.410000e+02  
defense_score     3.000000e+00  1.000000e+01  2.100000e+01  1.410000e+02  
game_id           4.012473e+08  4.014041e+08  4.015203e+08  4.015539e+08  
drive_id          4.014245e+09  4.012153e+10  4.014186e+10  4.015539e+10  
drive_number      6.000000e+00  1.200000e+01  1.900000e+01  4.900000e+01  
play_number       2.000000e+00  4.000000e+00  7.000000e+00  4.000000e+01  
period            2.000000e+00  2.000000e+00  3.000000e+00  8.000000e+00  
offense_timeouts  2.000000e+00  3.000000e+00  3.000000e+00  3.000000e+00  
defense_timeouts  2.000000e+00  3.000000e+00  3.000000e+00  3.000000e+00  
yard_line         3.200000e+01  5.100000e+01  7.000000e+01  1.220000e+02  
yards_to_goal     3.400000e+01  5.700000e+01  7.000000e+01  1.000000e+02  
down              1.000000e+00  2.000000e+00  3.000000e+00  5.000000e+00  
distance          6.000000e+00  1.000000e+01  1.000000e+01  8.700000e+01  
yards_gained      0.000000e+00  2.000000e+00  8.000000e+00  1.105000e+03  
ppa              -5.700000e-01 -1.100000e-01  8.100000e-01  7.410000e+00

In [8]:
df.head().T

0  \
id                                                 401110720101849902   
offense                                                          Duke   
offense_conference                                                ACC   
defense                                                       Alabama   
defense_conference                                                SEC   
home                                                          Alabama   
away                                                             Duke   
offense_score                                                       0   
defense_score                                                       0   
game_id                                                     401110720   
drive_id                                                   4011107201   
drive_number                                                        1   
play_number                                                         1   
period                                                              1   
clock                                   {'minutes': 15, 'seconds': 0}   
offense_timeouts                                                  3.0   
defense_timeouts                                                  3.0   
yard_line                                                          65   
yards_to_goal                                                      65   
down                                                                1   
distance                                                           10   
scoring                                                         False   
yards_gained                                                       22   
play_type                                                     Kickoff   
play_text           AJ Reed kickoff for 65 yds , Henry Ruggs III r...   
ppa                                                               NaN   
wallclock                                    2019-08-31T19:37:42.000Z   

                                                          1  \
id                                       401110720101855301   
offense                                             Alabama   
offense_conference                                      SEC   
defense                                                Duke   
defense_conference                                      ACC   
home                                                Alabama   
away                                                   Duke   
offense_score                                             0   
defense_score                                             0   
game_id                                           401110720   
drive_id                                         4011107201   
drive_number                                              1   
play_number                                               2   
period                                                    1   
clock                        {'minutes': 14, 'seconds': 46}   
offense_timeouts                                        3.0   
defense_timeouts                                        3.0   
yard_line                                                22   
yards_to_goal                                            78   
down                                                      1   
distance                                                 10   
scoring                                               False   
yards_gained                                              1   
play_type                                              Rush   
play_text           Jerome Ford run for 1 yd to the Alab 23   
ppa                                                -0.51422   
wallclock                          2019-08-31T19:38:08.000Z   

                                                                    2  \
id                                                 401110720101858401   
offense                                                       Alabama   
offense_conference                     

In [16]:
df['clock'][0]

"{'minutes': 15, 'seconds': 0}"

In [18]:
df['clock_minutes'] = df['clock'].apply(lambda x: eval(x)['minutes'] + eval(x)['seconds']/60)
df['clock_minutes'].value_counts(bins=10)

(-0.059000000000000004, 5.8]    347281
(5.8, 11.6]                     293302
(11.6, 17.4]                    229182
(46.4, 52.2]                         9
(40.6, 46.4]                         8
(17.4, 23.2]                         7
(34.8, 40.6]                         5
(52.2, 58.0]                         4
(29.0, 34.8]                         3
(23.2, 29.0]                         3
Name: count, dtype: int64

In [19]:
len(df[df['clock_minutes'] > 15])

39

In [22]:
len(df[df['game_id'].isin(df[df['clock_minutes'] > 15]['game_id'])])

570

Fine with dropping the games where any clock_minutes value exceeds 15 minutes. 570 plays is very small compared to the dataset size

In [23]:
df = df[~df['game_id'].isin(df[df['clock_minutes'] > 15]['game_id'])]

## Simple model

reducing to 
- `yards_to_goal`
- `down`
- `distance`
- `score_differential` -- which is `offense_score` - `defense_score`
- `time_remaining` -- a bit tricky with this dataset, but not impossible
    - create the `quarter` value
        - order by `game_id` and then `play_number`
        - the quarter gets determined by when the clock value resets to 15
        - From there, time remaining is just multiplication by the inverse of the quarter. clock * 4 for Q1, clock * 3 for Q2, etc. 
